src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala

```diff
@@ -1066,9 +1068,46 @@ abstract class BTypes {
    * @param inlineInfo    Information about this class for the inliner.
    */
   final case class ClassInfo(superClass: Option[ClassBType], interfaces: List[ClassBType], flags: Int,
-                             nestedClasses: List[ClassBType], nestedInfo: Option[NestedInfo],
+                             nestedClasses: Lazy[List[ClassBType]], nestedInfo: Lazy[Option[NestedInfo]],
                              inlineInfo: InlineInfo)
 
+  object Lazy {
+    def apply[T <: AnyRef](t: => T): Lazy[T] = new Lazy[T](() => t)
+  }
+
+  final class Lazy[T <: AnyRef](t: () => T) {
+    private var value: T = null.asInstanceOf[T]
+
+    private var function = {
+      val tt = t // prevent allocating a field for t
+      () => { value = tt() }
+    }
+
+    override def toString = if (value == null) "<?>" else value.toString
+
+    def onForce(f: T => Unit): Unit = {
+      if (value != null) f(value)
+      else {
+        val prev = function
+        function = () => {
+          prev()
+          f(value)
+        }
+      }
+    }
+
+    def force: T = {
+      if (value != null) value
+      else bTypes.synchronized {
```

Comment:
Well spotted! We'll need to use the same lock for both, as they wrap unknown callback code. This is similar the use of `gilSynchronized` in runtime reflection.